Support differential/velocity data in frame concatenate() function

adrn opened the pull request:
This PR does two things:

1. Fixes #7506 - the `concatenate()` function now properly handles differential data in frames.
2. Adds a new function to concatenate raw `Representation` instances (with or without attached differentials), and offloads most of the work here.

Previously, all of the work in `concatenate()` happened in the `SkyCoord` initializer.

eteq commented on CHANGES.rst:
```diff
@@ -243,6 +246,9 @@ astropy.convolution
 astropy.coordinates
 ^^^^^^^^^^^^^^^^^^^
 
+- Fixed ``astropy.coordinates.concatenate`` to also concatenate velocity (i.e.
```
(minor) Maybe "to include velocity in the concatenation" or similar?  Right now it sounds like it didn't concatenate *anything*

eteq commented on astropy/coordinates/funcs.py:
```diff
@@ -279,3 +254,120 @@ def get_constellation(coord, short_name=False, constellation_list='iau'):
+    # Check that all frames are equivalent
+    for sc in scs[1:]:
+        if not sc.is_equivalent_frame(scs[0]):
+            raise ValueError("All inputs must have equivalent frames: "
```
with a trivial `transform_to`, we could now make this put them all in the same frame... maybe that's a good enhancement (if you don't think it's a good idea it shouldn't block this PR, though)

adrn replied:
My gut reaction is to make the user decide, rather than to assume that the frame of the 1st object is special. So, something to consider for future but I'd rather not do that now.

eteq commented on astropy/coordinates/funcs.py:
```diff
@@ -279,3 +254,120 @@ def get_constellation(coord, short_name=False, constellation_list='iau'):
+    if getattr(coords, 'isscalar', False) or not isiterable(coords):
+        raise TypeError('The argument to concatenate must be iterable')
+
+    scs = [SkyCoord(coord) for coord in coords]
```
``copy=False`` for good measure here?  Or *maybe* for performance instead make the frame classes?

eteq commented on astropy/coordinates/funcs.py:
```diff
@@ -279,3 +254,120 @@ def get_constellation(coord, short_name=False, constellation_list='iau'):
+
+    return values
+
+def concatenate_representations(reps):
```
This needs to be in `__all__` to be in the public API, right?

adrn replied:
done

adrn commented:
> does this have better, worse, or same performance vs current master?

Sounds like I should add a benchmark ;)

adrn commented:
OK, I think I addressed the comments you left in here! 

On performance, this PR is ~5 times faster:

```python
c1 = coord.ICRS(ra=1*u.deg, dec=2*u.deg)
c2 = coord.ICRS(ra=np.random.random(10000)*u.deg, dec=np.random.random(10000)*u.deg)
```

This PR:
```python
%timeit coord.concatenate((c1, c1))
1.9 ms ± 408 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit coord.concatenate((c2, c2))
2.32 ms ± 388 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

master:
```python
%timeit coord.concatenate((c1, c1))
8.97 ms ± 340 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit coord.concatenate((c2, c2))
10.1 ms ± 910 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

eteq commented:
LGTM now - merging, since this is blocking other work.
